layers/ip6.go

```diff
@@ -280,165 +229,191 @@ func decodeIPv6(data []byte, p gopacket.PacketBuilder) error {
 	return p.NextDecoder(ip6.NextHeader)
 }
 
-func (i *IPv6HopByHop) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.SerializeOptions) error {
-	var bytes []byte
-	var err error
+type IPv6HeaderTLVOption interface {
```

Comment:
For gopacket, I've tried to err on the side of:

idiomatic Go - for exported functions, enums, etc.
very fast (possibly not idiomatic) Go - for actual implementation

Especially when it comes to packet decoding/encoding, where high throughput
is key, I think the ends justify the means when it comes to things like
avoiding mallocs inside the code.

I'm wondering if there might be a way to have our cake and eat it too...
maybe implement an interface, but only use it internally:

type fooIface interface {
  bar()
  blah()
}

func DoStuff(t *thing) {
  fiface := fooIface(t)
  // so long as fiface doesn't escape this function, no malloc.  And in
here, we can use the nice features of the interface.
}

Thoughts?

On Thu, Aug 27, 2015 at 6:55 PM, randstr <notifications@github.com> wrote:

> In layers/ip6.go
> <https://github.com/google/gopacket/pull/113#discussion_r38163245>:
>
> > @@ -280,165 +229,191 @@ func decodeIPv6(data []byte, p gopacket.PacketBuilder) error {
> >  	return p.NextDecoder(ip6.NextHeader)
> >  }
> >
> > -func (i *IPv6HopByHop) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.SerializeOptions) error {
> > -	var bytes []byte
> > -	var err error
> > +type IPv6HeaderTLVOption interface {
>
> The IPv6HeaderTLVOption interface doesn't have to be exported but it does
> allow users of the API to implement their own TLV types. Right? Though it
> is possible to use just the IPv6HeaderTLVOptionUnknown type.
>
> It doesn't have a DecodeFromBytes() method, though I suppose it could,
> with a suitable type code table. Unless I'm missing something.
>
> Dunno about the use of malloc. I feel like the code is more idiomatic Go
> and it shows. It doesn't treat IPv6HopByHopOption and IPv6DestinationOption
> as different objects. There are no BigEndian.Uint32() calls to set/get the
> values (for Pad and Jumbo at least). And the code to align and serialize
> the TLVs is also simpler and perhaps more elegant. It computes the length
> of the aligned/padded HopByHop/Destination header without a serialize dry
> run. Is that issue a showstopper?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/gopacket/pull/113/files#r38163245>.
>